app/demo-geant-integration/HepMC3Reader.cc

```diff
@@ -0,0 +1,162 @@
+
+#include <mutex>
+#include <G4SystemOfUnits.hh>
+#include <HepMC3/ReaderFactory.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+#include "GlobalSetup.hh"
+
+namespace
+{
+// Mutex visible to all threads
+static std::mutex mutex;
+} // namespace
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Singleton declaration.
+ */
+static HepMC3Reader* hepmc3_reader_singleton = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return non-owning pointer to a singleton.
+ */
+HepMC3Reader* HepMC3Reader::instance()
+{
+    {
+        mutex.lock();
+        if (!hepmc3_reader_singleton)
+        {
+            hepmc3_reader_singleton = new HepMC3Reader();
+        }
+        mutex.unlock();
+    }
+
+    CELER_LOG_LOCAL(status) << "Instancing HepMC3 singleton";
+    return hepmc3_reader_singleton;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add HepMC3 primaries to a Geant4 event. This function is called by
+ * `G4VUserPrimaryGeneratorAction::GeneratePrimaries`.
+ */
+void HepMC3Reader::GeneratePrimaryVertex(G4Event* g4_event)
+{
+    mutex.lock();
+
+    // Populate event_primaries
+    auto result = this->store_primaries();
+    CELER_ASSERT(result);
+
+    // Add primaries to event
+    for (const auto& primary : event_primaries_)
+    {
+        // TODO: Do we need to check if vertex is inside world volume?
+
+        // All primaries start at t = 0
+        auto g4_vtx = std::make_unique<G4PrimaryVertex>(
+            primary.vertex[0], primary.vertex[1], primary.vertex[2], 0);
+
+        g4_vtx->SetPrimary(new G4PrimaryParticle(primary.pdg,
+                                                 primary.momentum[0],
+                                                 primary.momentum[1],
+                                                 primary.momentum[2],
+                                                 primary.energy));
+
+        g4_event->AddPrimaryVertex(g4_vtx.release());
+    }
+    const int num_vertices = g4_event->GetNumberOfPrimaryVertex();
+    CELER_ENSURE(num_vertices);
+
+    mutex.unlock();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with HepMC3 input filename and store total number of events.
+ */
+HepMC3Reader::HepMC3Reader() : G4VPrimaryGenerator()
+{
+    std::string filename = GlobalSetup::Instance()->GetHepmc3File();
+    CELER_LOG_LOCAL(status) << "Constructing HepMC3 reader with " << filename;
+    input_file_ = HepMC3::deduce_reader(filename);
+
+    // Fetch total number of events
+    const auto temp_file = HepMC3::deduce_reader(filename);
+    num_events_          = -1;
+    while (!temp_file->failed())
+    {
+        // Count event and try to read the next
+        HepMC3::GenEvent gen_event;
+        temp_file->read_event(gen_event);
+        num_events_++;
+    }
```

Comment:
Why are the number of events needed? It seems like a lot of work to open and reread a file for something that's not (as far as I can see?) used anywhere.

Reply:
Sorry, I forgot to mention that in the PR description, and your review was fast. See my note below.

Reply:
Since we don't actually need to read the event here I think you could replace
```
HepMC3::GenEvent gen_event;
temp_file->read_event(gen_event);
```
with 
```
temp_file->skip(1);
```
Also, `num_events_` should start at 0.